README
Name

    Module::Provision - Create Perl distributions with VCS and selectable
    toolchain

Version

    This documents version v0.36.$Rev: 1 $ of Module::Provision

Synopsis

       # To reduce typing define a shell alias
       alias mp='module-provision --base ~/Projects'
    
       # Create a new distribution in your Projects directory with Git VCS
       mp dist Foo::Bar 'Optional one line abstract'
    
       # Add another module
       cd ~/Projects/Foo-Bar
       mp module Foo::Bat 'Optional one line abstract'
    
       # Add a program to the bin directory
       mp program bar-cli 'Optional one line abstract'
    
       # Add another test script
       mp test 11another-one.t
    
       # Edit the project master file
       mp -q edit_project
    
       # Regenerate meta data files
       mp metadata
    
       # Update the version numbers of the project files
       mp update_version 0.1 0.2
    
       # Stateful setting of the current working branch
       mp set_branch <branch_name>
    
       # Command line help
       mp -? | -H | -h [sub-command] | list_methods | dump_self

Description

    Module::Provision is used to create a skeletal CPAN distribution,
    including basic builder scripts, tests, documentation, and module code.
    It creates a VCS repository and, with Git as the VCS, installs some
    hooks that mimic the RCS Revision keyword expansion

    On first use the directory ~/.module_provision is created and populated
    with templates and an index file index.json. The author name, id, and
    email are derived from the system (the environment variables AUTHOR and
    EMAIL take precedence). They can be overridden by the values in the
    configuration file ~/.module_provision/module_provision.json

    If the default builder (MB) is used, then the project file Build.PL
    loads inc::Bob which instantiates an inline subclass of Module::Build.
    The code for the subclass is in inc::SubClass. The file
    inc::CPANTesting allows for fine grained control over which tests are
    run by which CPAN Testing smokers

    The default builder used by the create distribution method can be
    changed from the command line or from the configuration file

    If the Git VCS is used precommit and commit-msg hooks are installed.
    The precommit hook will expand the RCS Revision keyword in files on the
    master branch if the file .distribution_name.rev exists in the parent
    of the working tree. The precommit hook will also update the version
    number and date/time stamp in the change log (Changes). The commit-msg
    hook will extract the first comment line from the change log and use it
    as the commit message header. The remainder of the commit message (if
    any) is used as the commit message body. This means that so long as one
    detail line is added to the change log no other commit message text is
    required. The following makes for a suitable git log alias:

       alias gl='git log -5 --pretty=format:"%h %ci %s" | \
          cut -d" " -f1-3,5- | cut -c1-79'

    The default VCS used by the create distribution methods can be changed
    from the command line or from the configuration file

    The templates contain comment lines like:

       # @(#)Ident: Provision.pm 2013-04-15 13:52 pjf ;

    These are expanded automatically by Emacs using this Lisp code:

       (setq time-stamp-active     t
             time-stamp-line-limit 10
             time-stamp-format     " %f %04y-%02m-%02d %02H:%02M %u "
             time-stamp-start      "Ident:"
             time-stamp-time-zone  "UTC"
             time-stamp-end        ";")

    The alias:

       alias ident='ack "@\(#\)"'

    uses the App::Ack program to implement the old SYSV R4 ident command

    The templates for the project files dist.ini, Build.PL, and Makefile.PL
    contain the following comments which are interpreted by Emacs:

       # Local Variables:
       # mode: perl
       # eval: (load-project-state "[% appdir %]")
       # tab-title: [% project %]
       # tab-width: 3
       # End:

    Perl mode is preferred over C-Perl mode since the former has better
    syntax highlighting. Tabs are expanded to three spaces. The tab-title
    variable is used by Yakuake::Sessions to set the tab title for the
    terminal emulator. The load-project-state Lisp looks like this:

       (defun load-project-state (state-file) "Recovers the TinyDesk state from file"
          (let ((session-path (concat "~/.emacs.d/config/state." state-file)))
             (if (file-exists-p session-path) (tinydesk-recover-state session-path)
                (message (concat "Not found: " state-file)))))

    It assumes that the TinyDesk state file containing the list of files to
    edit for the project has been saved in ~/.emacs.d/config/state.[%
    appdir %]. To work on a project; change directory to the working copy,
    edit the project file Build.PL with Emacs, this will load all of the
    other files in the project into separate buffers displaying each in the
    tab bar. This Lisp code will load TinyDesk and turn tab bar mode on
    whenever a Perl file is edited:

       (add-hook 'perl-mode-hook
                 '(lambda ()
                    (require 'fic-mode) (turn-on-fic-mode) (diminish 'fic-mode nil)
                    (require 'psvn) (require 'tinydesk) (tabbar-mode t)
                    (require 'tinyperl) (diminish 'tinyperl-mode nil)))

    This Lisp code will do likewise when a dist.ini file is edited:

       (add-hook 'conf-windows-mode-hook
                 '(lambda ()
                    (require 'tinydesk) (tabbar-mode t)))

Configuration and Environment

    The configuration file defaults to
    ~/.module_provision/module_provision.json. All of the attributes listed
    in Module::Provision::Config can be set from the configuration file in
    addition to the attributes listed in Class::Usul::Config::Programs and
    Class::Usul::Config. A typical file looks like;

       {
          "author": "<first_name> <last_name>",
          "author_email": "<userid>@example.com",
          "author_id": "<userid>",
          "base": "/home/<userid>/Projects",
          "doc_title": "Perl",
          "editor": "emacs",
          "home_page": "http://www.example.com"
       }

    Creating logs and tmp directories in ~/.module_provision will cause the
    log and temporary files to use them instead of /tmp

    Extends Module::Provision::Base. Applies these traits; AddingFiles,
    CreatingDistributions, Rendering, UpdatingContent, and VCS

    This class defines no attributes

Subroutines/Methods

 cpan_upload

       module-provision cpan_upload <optional_version_number>

    By default uploads the projects current distribution to CPAN

 delete_cpan_files

       module-provision delete_cpan_files v0.1.1

    Deletes a specified version of the projects distributions from CPAN

 dist

       module-provision dist Foo::Bar <'Optional one line abstract'>

    Create a new distribution specified by the module name on the command
    line

 dump_stash

       module-provision dump_stash

    Dump the hash ref used to render a template

 edit_project

       module-provision -q edit_project

    Edit the project file (one of; dist.ini, Build.PL, or Makefile.PL) in
    the project directory. The editor defaults to emacs but can be set on
    the command line, e.g -o editor=vim

 metadata

       module-provision metadata

    Generates the distribution metadata files

 init_templates

       module-provision init_templates

    Initialise the .module_provision directory and create the index.json
    file

 module

       module-provision module Foo::Bat <'Optional one line abstract'>

    Creates a new module specified by the class name on the command line

 program

       module-provision program bar-cli <'Optional one line abstract'>

    Creates a new program specified by the program name on the command line

 prereq_diffs

       module-provision prereq_diffs

    Displays a report showing which pre-requisite modules should be added
    to, removed from, or updated in the project file

 prove

       module-provision prove

    Runs the projects tests

 select_project

       cd $(module_provision -q select_project 3>&1 1>/dev/tty 2>/dev/null)

    Displays a list of available projects. Calls edit_project on the
    selected option

 set_branch

       module-provision set_branch <branch_name>

    Persistently sets the branch name used on this project. If branch_name
    is omitted defaults to the branch name appropriate for the VCS being
    used. Edits the currently selected editor's state file for the project
    to reflect the changing pathnames

 set_cpan_password

       module-provision set_cpan_password <your_PAUSE_server_password>

    Sets the password used to connect to the PAUSE server. Once used the
    command line program cpan-upload will not work since it cannot decrypt
    the password in the configuration file ~/.pause

 show_tab_title

       module-provision -q show_tab_title

    Print the tab title for the current project. Can be used like this;

       alias ep='mp -q edit_project ; \
          yakuake_session set_tab_title_for_project $(mp -q show_tab_title)'

 test

       module-provision test 11another-one.t

    Creates a new test specified by the test file name on the command line

 update_copyright_year

       module-provision update_copyright_year 2013 2014

    Substitutes the existing copyright year for the new copyright year in
    all files in the MANIFEST

 update_version

       module-provision update_version 0.1 0.2

    Substitutes the existing version number for the new version number in
    all files in the MANIFEST. Prompts for the major/minor and bump if the
    version numbers are not provided

Diagnostics

    Add -D to command line to turn on debug output

Dependencies

    Module::Provision::Base

    Module::Provision::TraitFor::AddingFiles

    Module::Provision::TraitFor::CreatingDistributions

    Module::Provision::TraitFor::Rendering

    Module::Provision::TraitFor::UpdatingContent

    Module::Provision::TraitFor::VCS

    Moo

Incompatibilities

    There are no known incompatibilities in this module

Bugs and Limitations

    There are no known bugs in this module. Please report problems to
    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Module-Provision. Source code
    is on Github git://github.com/pjfl/Module-Provision.git. Patches and
    pull requests are welcome

Acknowledgements

    Larry Wall - For the Perl programming language

    Module::Starter - For some of the documentation and tests

Author

    Peter Flanigan, <pjfl@cpan.org>

License and Copyright

    Copyright (c) 2013 Peter Flanigan. All rights reserved

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic

    This program is distributed in the hope that it will be useful, but
    WITHOUT WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE



Changes
Revision history for Module::Provision

0.36.1    2014-10-28 13:45:12
          - Fixed spelling

0.35.11   2014-10-28 13:29:54
          - Added release method to ::VCS
          - Added t::boilerplate.pm
          - Add .pm files to prereq_diffs for build_requires
          - Ignore t/* files in prereq_diffs for requires
          - Replaced namespace::sweep with namespace::autoclean
          - Added fury badge
          - Better parsing of module name from project file
          - Validate args in call to update_copyright_year

0.34.1    2014-05-19 12:45:10
          - c6a96d00-df30-11e3-84c4-fc77f9652e90 Test::Requires Hash::MoreUtils

0.33.1    2014-05-15 21:54:47
          - Removed dependency on File::Slurp. Not needed since
            D:Z:P:ManifestInRoot became a seperate dist

0.32.1    2014-05-01 15:32:09
          - Removed Ident and VERSION lines

0.31.2    2014-01-24 21:04:29
          - Updated dependencies
          - Select project does better job of picking the project dir

0.30.1    2014-01-15 18:05:13
          - Using Unexpected exception classes
          - Stopped using C::U::File::io
          - Apply C::U::T::MetaData for get_package_meta method
          - Stopped localizing error messages
          - Added update_file method to copy from template dir

0.29.1    2013-12-12 21:26:25
          - Remove prereqs below min_perl_ver

0.28.1    2013-12-12 14:16:26
          - Allow CPAN Testing to smoke below min_perl_ver

0.27.1    2013-12-11 17:23:36
          - Dropped create_readme_md from share/Build.PL
            Closes RT#91320
          - Tweaked boilerplate in share/*.t
          - Simplified share/DZ_Build.PL
          - Use DZ::P::AbstractFromPOD
          - Use DZ::P::LicenseFromModule
          - Fixed fd3 usage in select_project

0.26.1    2013-11-25 21:28:18
          - Dropped DZ::P::MarkdownInRoot from dist.ini
          - Added select_project
          - Added ::Config::test_env_vars

0.25.1    2013-11-23 13:35:51
          - Dropped dependency on MooX::Options
          - Made ensure_class_loaded into a function

0.24.1    2013-09-16 21:22:24
          - Stopped re interpretaion of homedir in appldir builder

0.23.1    2013-09-14 10:37:11
          - Bumped Moo dependency closes RT#87575

0.22.1    2013-08-21 13:54:58
          - Bumped version to fix indexing

0.21.1    2013-08-21 13:25:21
          - Renamed Github repository to p5-module-provision

0.20.2    2013-08-19 14:59:28
          - Filter Dzil project file from prereq report
          - Fixed prereq report with Dzil
          - Bumped dependencies

0.19.5    2013-08-17 16:37:36
          - Switched to Dist::Zilla

0.18.4    2013-08-10 22:17:28
          - Updated share/dist.ini. Added share/DZ_Build.PL
          - Stopping testing on windows until RT#87575 is fixed
            febd397b-6bf5-1014-b508-69147d50f5d1
          - Toolchain update. Attempt to filter out residual noise
          - Misconfigured smoker 3c25f6a5-6c83-1014-9715-b1a9b77d427c
            c.f. df2446e9-6c38-1014-8bcf-004655012b15. Cannot see error
            because of RT#87575. Toolchain updated
          - Moved D::Z::P::ManifestInRoot to own dist


Build.PL
# @(#)$Ident: DZ_Build.PL 2013-12-14 14:19 pjf ;

use strict;
use warnings;

use Module::Build 0.4004;

sub whimper { print {*STDOUT} $_[ 0 ]."\n"; exit 0 }

my $perl_ver = 5.010001;
my $name     = 'Module-Provision'; $name =~ s{ - }{::}gmx;
my $prereqs  = {
  'build_requires' => {
    'Module::Build' => '0.4004',
    'Test::Requires' => '0.06',
    'version' => '0.88'
  },
  'configure_requires' => {
    'Module::Build' => '0.4004',
    'version' => '0.88'
  },
  'requires' => {
    'CPAN::Uploader' => '0.103004',
    'Class::Null' => '2.101420',
    'Class::Usul' => 'v0.50.0',
    'Config::Tiny' => '2.20',
    'Date::Format' => '2.24',
    'Dist::Zilla' => '4.300033',
    'Dist::Zilla::Plugin::AbstractFromPOD' => 'v0.2.0',
    'Dist::Zilla::Plugin::LicenseFromModule' => '0.03',
    'Dist::Zilla::Plugin::ManifestInRoot' => 'v0.1.0',
    'Dist::Zilla::Plugin::Meta::Dynamic::Config' => '0.04',
    'Dist::Zilla::Plugin::ModuleBuild::Custom' => '4.16',
    'Dist::Zilla::Plugin::ReadmeAnyFromPod' => '0.141760',
    'Dist::Zilla::Plugin::Signature' => '1.100930',
    'Dist::Zilla::Plugin::VersionFromModule' => '0.08',
    'File::DataClass' => 'v0.49.0',
    'File::ShareDir' => '1.102',
    'HTTP::Message' => '6.06',
    'LWP' => '6.08',
    'LWP::Protocol::https' => '6.03',
    'Module::Install' => '1.12',
    'Module::Install::AutoLicense' => '0.08',
    'Module::Install::ReadmeFromPod' => '0.20',
    'Module::Install::ReadmeMarkdownFromPod' => '0.03',
    'Module::Metadata' => '1.000011',
    'Moo' => '1.006000',
    'Perl::Version' => '1.013',
    'Pod::Eventual' => '0.094001',
    'Software::License' => '0.103010',
    'Template' => '2.26',
    'Type::Tiny' => '1.000004',
    'Unexpected' => 'v0.34.0',
    'local::lib' => '2.000014',
    'namespace::autoclean' => '0.20',
    'perl' => '5.010001'
  },
};

if ($] < $perl_ver) {
   $ENV{AUTOMATED_TESTING}
      or $ENV{PERL_CPAN_REPORTER_CONFIG}
      or $ENV{PERL5_CPANPLUS_IS_VERSION}
      or whimper "Minimum required Perl version is ${perl_ver}";
   $prereqs->{build_requires} = { %{ $prereqs->{configure_requires} } };
   $prereqs->{build_requires}->{ 'Test::Requires' } = 0.06;
   delete $prereqs->{requires};
}

Module::Build->new(
  'module_name' => $name,
  'notes'       => { min_perl_version => $perl_ver },
  %{ $prereqs },
  'share_dir' => {
    'dist' => 'share'
  },
)->create_build_script;

exit 0;
